ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Time;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * DefaultCompactionPolicy calculates the cost benefit ratio for every potential candidate to compact and finds the
+ * best one to be compacted to be returned for {@link #getCompactionDetails(long, long, long, long, List, BlobStoreStats)}
+ */
+class DefaultCompactionPolicy implements CompactionPolicy {
+
+  private final Time time;
+  private final StoreConfig storeConfig;
+  private final long messageRetentionTimeInMs;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  DefaultCompactionPolicy(StoreConfig storeConfig, Time time) {
+    this.storeConfig = storeConfig;
+    this.time = time;
+    this.messageRetentionTimeInMs = storeConfig.storeDeletedMessageRetentionDays * Time.SecsPerDay * Time.MsPerSec;
+  }
+
+  @Override
+  public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapacity, long segmentCapacity,
+      long segmentHeaderSize, List<String> logSegmentsNotInJournal, BlobStoreStats blobStoreStats)
+      throws StoreException {
+    CompactionDetails details = null;
+    if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
+      if (logSegmentsNotInJournal != null) {
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
+
+        Iterator<Map.Entry<String, Long>> validEntryIterator =
+            validDataPerLogSegments.getSecond().entrySet().iterator();
+        while (validEntryIterator.hasNext()) {
```

Comment:
is this loop required at all? What if we took a submap starting from the beginning to the last element in `logSegmentsNotInJournal`?